handler/verifiable/flow_preauthcode_token.go

```diff
@@ -0,0 +1,195 @@
+
+package verifiable
+
+import (
+	"context"
+	"time"
+
+	"github.com/ory/fosite"
+	"github.com/ory/fosite/handler/oauth2"
+	"github.com/ory/fosite/storage"
+	"github.com/ory/x/errorsx"
+	"github.com/pkg/errors"
+)
+
+var _ fosite.TokenEndpointHandler = (*PreAuthorizeCodeTokenHandler)(nil)
+
+type PreAuthorizeCodeTokenHandler struct {
+	AuthorizeCodeStrategy  oauth2.AuthorizeCodeStrategy
+	AccessTokenStrategy    oauth2.AccessTokenStrategy
+	RefreshTokenStrategy   oauth2.RefreshTokenStrategy
+	TokenRevocationStorage oauth2.TokenRevocationStorage
+	Storage                Storage
+	Config                 interface {
+		fosite.AccessTokenLifespanProvider
+		fosite.RefreshTokenLifespanProvider
+		fosite.ScopeStrategyProvider
+		fosite.AudienceStrategyProvider
+		fosite.RefreshTokenScopesProvider
+	}
+}
+
+// HandleTokenEndpointRequest handles an authorize request. If the handler is not responsible for handling
+// the request, this method should return ErrUnknownRequest and otherwise handle the request.
+func (c *PreAuthorizeCodeTokenHandler) HandleTokenEndpointRequest(ctx context.Context, request fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(ctx, request) {
+		return errorsx.WithStack(errorsx.WithStack(fosite.ErrUnknownRequest))
+	}
+
+	client := request.GetClient()
+
+	// Check whether client is allowed to use pre-authorize grant type
+	if !client.GetGrantTypes().Has(string(fosite.GrantTypePreAuthorizeCode)) {
+		return errors.WithStack(fosite.ErrUnauthorizedClient.WithHintf(
+			"The OAuth 2.0 Client is not allowed to use authorization grant \"%s\".", fosite.GrantTypePreAuthorizeCode))
+	}
+
+	// Check scope requested
+	for _, scope := range request.GetRequestedScopes() {
+		if !c.Config.GetScopeStrategy(ctx)(client.GetScopes(), scope) {
+			return errors.WithStack(fosite.ErrInvalidScope.WithHintf("The OAuth 2.0 Client is not allowed to request scope '%s'.", scope))
+		}
+	}
+
+	// Check audience requested
+	if err := c.Config.GetAudienceStrategy(ctx)(client.GetAudience(), request.GetRequestedAudience()); err != nil {
+		return err
+	}
+
+	// load the session based on incoming pre-authorize_code request parameter
+	code := request.GetRequestForm().Get("pre-authorized_code")
+
+	signature := c.AuthorizeCodeStrategy.AuthorizeCodeSignature(ctx, code)
+	preAuthRequest, err := c.Storage.GetPreAuthorizeSession(ctx, signature, request.GetSession())
+	if err != nil && errors.Is(err, fosite.ErrNotFound) {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.WithHint("Invalid Pre-Authorized Code or the Pre-Authorized Code has expired."))
+	} else if err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	// The authorization server MUST verify that the pre-authorization code is valid
+	// This needs to happen after store retrieval for the session to be hydrated properly
+	if err := c.AuthorizeCodeStrategy.ValidateAuthorizeCode(ctx, request, code); err != nil {
```

Comment:
Should we be introducing a new function specifically for pre-authorized code to make sure that the pre auth code isn't misused as the auth code?

Reply:
yeah, make sense. Ok I'll separate it.

Reply:
I'm reusing the generation and validation of the authorization code string only. Not the storing part. So, I think it should be ok.